### PR TITLE
Add shift for location of camera.

### LINKED ISSUE #14846: Mismatch between camera target and device location when in `TRACKING` (in China)
This ticket is to report a mismatch between the camera target and the actual device location, when the Mapbox China Plugin for Android is being used and the `LocationComponent` is in `TRACKING` camera mode. 

### Steps to reproduce

 1. Create a barebones Android app with a Mapbox map **and** [the Mapbox China Plugin for Android](https://docs.mapbox.com/android/plugins/overview/china/#install-the-china-plugin) (requires a special Mapbox China token). 
2. Also set up [the `LocationComponent`](https://docs.mapbox.com/android/maps/overview/location-component/) with [the `TRACKING` camera mode](https://docs.mapbox.com/android/maps/overview/location-component/#cameramode).

### Expected behavior

The camera stays directly centered on the device location puck.

### Actual behavior

The camera target location has a mismatch with the device's location (puck).

### Configuration

**Android versions:**  Pie
**Device models:**  OnePlus 5
**Mapbox SDK versions:**  8.0.0

### PR DESCRIPTION
**Pull request by Chaoba**
Resolves #14846 
Shift is not used on camera, so  it will show wrong location when using location shift.
In this PR, we implement shift in `CameraPosition.Builder().build()` method for the origin location. Also add test demo in test app.

### CODE REVIEW

**Comment by Chaoba** [comment]
@LukasPaczos If make shift in `CameraPosition#Builder` it could be shifted twice or even multiple times since we don't know the passed in param `target` is shifted or not.

Btw, we can't unshift location in china plugin.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/camera/CameraPosition.java [hunk 1]
```diff
@@ -168,7 +170,7 @@ public int hashCode() {
    * Builder for composing CameraPosition objects.
    */
   public static final class Builder {
-
+    private boolean needShift = true;
```
Comment (LukasPaczos) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/camera/CameraPosition.java]: `needShift` should be `false` by default and only set to `true` if  `Builder#target` is called.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/camera/CameraPosition.java [hunk 2]
```diff
@@ -312,6 +318,10 @@ public Builder zoom(@FloatRange(from = MapboxConstants.MINIMUM_ZOOM,
      * @return CameraPosition
      */
     public CameraPosition build() {
+      if (needShift && target != null) {
+        Point point = Point.fromLngLat(target.getLongitude(), target.getLatitude());
```
Comment (LukasPaczos) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/camera/CameraPosition.java]: You can use `CoordinateShifterManager.getCoordinateShifter().shiftLonLat()` to make the operation less expensive. We should make it possible to check if a custom shifter is set though, to minimize the memory overhead even more - https://github.com/mapbox/mapbox-java/issues/1057.

**Comment by LukasPaczos** [comment]
Thanks for the updates @Chaoba. I revisited the setup and here are a couple of thoughts:
- Currently, this simple assertion is going to fail, which feels unacceptable to me:
```
LatLng latLng = new LatLng(50, 16);
cameraBuilder.target(latLng);
CameraPosition position = cameraBuilder.build();
assertEquals(latLng, position.target);
```

- I looked through the code, and in most of the places, the geometry objects are passed to the `NativeMapView` and then to core.
- How about we introduce a new `shift` method in the classes found in the `com.mapbox.mapboxsdk.geometry` that is going to return a new object with shifted geometries, without changing the original? We could call this method in the `NativeMapView`, right before it's passed to core.
- One missing piece is the `MapSnapshotter#nativeInitialize` which passes the whole camera object directly and would have to be shifted there.
- Another arising problem is round-tripping - our use-case doesn't expect it, so below assertion is always going to fail:
```
map.setCameraPosition(value);
assertTrue(value, map.getCameraPosition);
```
- However, for other applications (are there any?) we could still include the unshift calls in constructors that are called from core.

Let me know what's your take @Chaoba. Any other thoughts @mapbox/maps-android?

**Comment by Chaoba** [comment]
@LukasPaczos Thanks for your thoughts. If we add a `shift` method, then in which situations should we call it and which situations should we use the origin one? How to prevent shift multi times?

After due consideration I think we shift location in `CameraPosition` is not a good idea, we should shift location out of it.

**Comment by LukasPaczos** [comment]
> in which situations should we call it and which situations should we use the origin one? How to prevent shift multi times?

We should call it right before geometry is passed to the core (for example in the `NativeMapView`). This way C++ objects will be shifted and Java objects will always be unshifted. I don't think we can run into double shifting this way unless we take into account round-tripping, but since our use-case doesn't provide a reverse algorithm, there's nothing we can do about it.

**Comment by tobrun** [comment]
We are removing platform code from gl-native in https://github.com/mapbox/mapbox-gl-native/pull/15970. If you want to see this PR merged, it will require re-implementation in https://github.com/mapbox/mapbox-gl-native-android.

**Pull request closed without merging**